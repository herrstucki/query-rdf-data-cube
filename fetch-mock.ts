import fetch from "fetch-vcr";

fetch.configure({
  fixturePath: "./_fixtures",
  mode: "cache",
});

export default fetch;
