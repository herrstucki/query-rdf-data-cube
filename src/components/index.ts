import Attribute from "./attribute";
import Component from "./component";
import Dimension from "./dimension";
import Measure from "./measure";

export default Component;
export {
  Dimension,
  Attribute,
  Measure,
};
