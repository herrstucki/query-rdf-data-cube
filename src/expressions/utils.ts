// tslint:disable: max-classes-per-file
import { blankNode, defaultGraph, literal, namedNode, variable } from "@rdfjs/data-model";
import { Literal, Term } from "rdf-js";
import { IExpr } from "./iexpr";

/**
 * @ignore
 */
export type IntoExpr = Term | string | number | IExpr;

/**
 * @ignore
 */
export class TermExpr implements IExpr {
  public term: Term;

  constructor(term: Term) {
    this.term = term;
  }

  public resolve(): TermExpr {
    return this;
  }
}

/**
 * @ignore
 */
export class ArrayExpr implements IExpr {
  public xs: Term[];

  constructor(xs: Term[]) {
    this.xs = xs;
  }

  public resolve(mapping: Map<string, string>) {
    return this;
  }
}

/**
 * @ignore
 */
export function isTerm(term: any): term is Term {
  return (
    term instanceof namedNode("").constructor ||
    term instanceof literal("").constructor ||
    term instanceof variable("").constructor ||
    term instanceof blankNode().constructor ||
    term instanceof defaultGraph().constructor
  );
}

/**
 * @ignore
 */
export function isLiteral(term: any): term is Literal {
  return (term instanceof literal("").constructor);
}
