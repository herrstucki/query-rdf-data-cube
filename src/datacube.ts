import { namedNode, variable } from "@rdfjs/data-model";
import namespace from "@rdfjs/namespace";
import { Literal, NamedNode, Term } from "rdf-js";
import { Generator as SparqlGenerator } from "sparqljs";
import { Attribute, Component, Dimension, Measure } from "./components";
import { EntryPointOptions } from "./entrypoint";
import { Query, QueryOptions } from "./query";
import { generateLangCoalesce, generateLangOptionals, prefixes } from "./queryutils";
import { SparqlFetcher } from "./sparqlfetcher";
import { SelectQuery } from "./sparqljs";

/**
 * @ignore
 */
type ComponentsCache = {
  attributes: Map<string, Attribute>,
  dimensions: Map<string, Dimension>,
  measures: Map<string, Measure>,
};
/**
 * @ignore
 */
type GroupedComponents = { kind: Term, iri: Term, labels: Label[] };

type SerializedDataCube = {
  endpoint: string,
  iri: string,
  graphIri: string,
  labels: Label[],
  languages: string[],
  components: {
    dimensions: string[],
    measures: string[],
    attributes: string[],
  },
};

/**
 * Represents an RDF
 * [language-tagged string](https://www.w3.org/TR/rdf11-concepts/#dfn-language-tagged-string).
 */
export interface Label {
  /**
   * String value
   */
  value: string;
  /**
   * Language tag such as "en", "fr", "de".
   * See [Tags for Identifying Languages](https://tools.ietf.org/html/bcp47).
   */
  language: string;
}

export interface DataCubeOptions extends EntryPointOptions {
  iri: NamedNode;
  labels?: Label[];
  graphIri: NamedNode;
}

const cube = namespace("http://purl.org/linked-data/cube#");

/**
 * @class DataCube
 */
export class DataCube {
  /**
   * Deserializes a [[DataCube]] from JSON generated by [[toJSON]]
   */
  public static fromJSON(json: string): DataCube {
    const obj: SerializedDataCube = JSON.parse(json);
    const dataCube = new DataCube(obj.endpoint, {
      iri: namedNode(obj.iri),
      graphIri: namedNode(obj.graphIri),
      labels: obj.labels,
      languages: obj.languages,
    });
    ["dimensions", "measures", "attributes"].forEach((componentTypes) => {
      dataCube.cachedComponents[componentTypes] = obj.components[componentTypes]
        .map(Component.fromJSON)
        .reduce((cache: Map<string, Component>, component: Component) => {
          cache.set(component.iri.value, component);
          return cache;
        }, new Map());
    });
    return dataCube;
  }

  public labels: Label[];
  public iri: string;
  public endpoint: string;
  public graphIri?: string;
  private languages: string[];
  private fetcher: SparqlFetcher;
  private componentsLoaded: boolean = false;
  private cachedComponents: ComponentsCache;

   /**
    * @param endpoint SPARQL endpoint where the DataCube lives.
    * @param options Additional info about the DataCube.
    * @param options.iri The IRI of the DataCube.
    * @param options.graphIri The IRI of the graph from which the data will be fetched.
    * @param options.labels (Optional) A list of labels for the DataCube in the following form:
    * `[ { value: "Something", language: "en" }, { value: "Etwas", language: "de" }, â€¦ ]`
    * @param options.languages Languages in which to get the labels, by priority, e.g. `["de", "en"]`.
    */
  constructor(
    endpoint: string,
    options: DataCubeOptions,
  ) {
    const { iri, labels, graphIri } = options;
    this.fetcher = new SparqlFetcher(endpoint);
    this.endpoint = endpoint;
    this.iri = iri.value;
    this.graphIri = graphIri.value;
    this.labels = labels || [];
    this.languages = options.languages || [];
    this.cachedComponents = {
      dimensions: new Map(),
      measures: new Map(),
      attributes: new Map(),
    };
  }

  /**
   * Serializes a [[DataCube]] to JSON in a way that makes it deserializable
   * by calling [[fromJSON]]
   */
  public toJSON(): string {
    const dimensions = Array.from(this.cachedComponents.dimensions.values())
      .map((component) => component.toJSON());
    const measures = Array.from(this.cachedComponents.measures.values())
      .map((component) => component.toJSON());
    const attributes = Array.from(this.cachedComponents.attributes.values())
      .map((component) => component.toJSON());
    const obj: SerializedDataCube = {
      endpoint: this.endpoint,
      iri: this.iri,
      graphIri: this.graphIri,
      labels: this.labels,
      languages: this.languages,
      components: {
        dimensions,
        measures,
        attributes,
      },
    };
    return JSON.stringify(obj);
  }

  /**
   * Fetch all [[Attribute]]s from the [[DataCube]].
   */
  public async attributes(): Promise<Attribute[]> {
    await this.components();
    return Array.from(this.cachedComponents.attributes.values());
  }

  /**
   * Fetch all [[Dimension]]s from the [[DataCube]].
   */
  public async dimensions(): Promise<Dimension[]> {
    await this.components();
    return Array.from(this.cachedComponents.dimensions.values());
  }

  /**
   * Fetch all [[Measure]]s from the [[DataCube]].
   */
  public async measures(): Promise<Measure[]> {
    await this.components();
    return Array.from(this.cachedComponents.measures.values());
  }

  /**
   * Start a new query on the DataCube.
   */
  public query(opts: QueryOptions = {}): Query {
    if (!opts.languages) {
      opts.languages = this.languages;
    }
    return new Query(this, opts);
  }

  public async componentValues(component: Component): Promise<Array<{label: Literal, value: NamedNode}>> {
    if (!component || !component.componentType) {
      throw new Error(`dataCube#componentValues expects valid component, got ${component} instead`);
    }
    const binding = variable("value");
    const labelBinding = variable("label");
    const observation = variable("observation");

    const query: SelectQuery = {
      prefixes,
      queryType: "SELECT",
      variables: [binding, labelBinding],
      distinct: true,
      from: { default: [namedNode(this.graphIri)], named: [] },
      where: [
        {
          type: "bgp",
          triples: [
            {
              subject: observation,
              predicate: namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"),
              object: cube("Observation"),
            },
            {
              subject: observation,
              predicate: component.iri,
              object: binding,
            },
            {
              subject: observation,
              predicate: cube("dataSet"),
              object: namedNode(this.iri),
            },
          ],
        },
        ...generateLangOptionals(binding, labelBinding, this.languages),
        generateLangCoalesce(labelBinding, this.languages),
      ],
      type: "query",
    };

    const generator = new SparqlGenerator({ allPrefixes: true });
    const sparql = generator.stringify(query);
    return await this.fetcher.select(sparql);
  }

  public async componentMinMax(component: Component): Promise<{min: Literal, max: Literal}|null> {
    if (!component || !component.componentType) {
      throw new Error(`dataCube#componentMinMax expects valid component, got ${component} instead`);
    }
    const binding = variable("value");
    const observation = variable("observation");

    const query: SelectQuery = {
      prefixes,
      queryType: "SELECT",
      variables: [
        {
          expression: {
            expression: binding,
            type: "aggregate",
            aggregation: "min",
            distinct: false,
          },
          variable: variable("min"),
        },
        {
          expression: {
            expression: binding,
            type: "aggregate",
            aggregation: "max",
            distinct: false,
          },
          variable: variable("max"),
        },
      ],
      from: { default: [namedNode(this.graphIri)], named: [] },
      where: [
        {
          type: "bgp",
          triples: [
            {
              subject: observation,
              predicate: namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"),
              object: cube("Observation"),
            },
            {
              subject: observation,
              predicate: component.iri,
              object: binding,
            },
            {
              subject: observation,
              predicate: cube("dataSet"),
              object: namedNode(this.iri),
            },
          ],
        },
        {
          type: "filter",
          expression: {
            type: "operation",
            operator: "isliteral",
            args: [binding],
          },
        },
      ],
      type: "query",
    };

    const generator = new SparqlGenerator({ allPrefixes: true });
    const sparql = generator.stringify(query);
    const results = await this.fetcher.select(sparql);
    if (results.length) {
      return results[0];
    }
    return null;
  }

  private async components() {
    if (this.componentsLoaded) {
      return;
    }

    const binding = variable("iri");
    const labelBinding = variable("label");

    const query: SelectQuery = {
      prefixes,
      queryType: "SELECT",
      variables: [
        binding,
        variable("kind"),
        labelBinding,
      ],
      from: { default: [namedNode(this.graphIri)], named: [] },
      where: [
        {
          type: "bgp",
          triples: [
            {
              subject: namedNode(this.iri),
              predicate: namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"),
              object: cube("DataSet"),
            },
            {
              subject: namedNode(this.iri),
              predicate: {
                type: "path",
                pathType: "/",
                items: [
                  cube("structure"),
                  cube("component"),
                ],
              },
              object: variable("componentSpec"),
            },
            {
              subject: variable("componentSpec"),
              predicate: variable("kind"),
              object: binding,
            },
          ],
        },
        {
          type: "filter",
          expression: {
            type: "operation",
            operator: "in",
            args: [
              variable("kind"),
              [
                cube("attribute"),
                cube("dimension"),
                cube("measure"),
              ],
            ],
          },
        },
        ...generateLangOptionals(binding, labelBinding, this.languages),
        generateLangCoalesce(labelBinding, this.languages),
      ],
      type: "query",
    };

    const generator = new SparqlGenerator({ allPrefixes: true });
    const sparql = generator.stringify(query);

    const components = await this.fetcher.select(sparql);
    const componentsByIri = components.reduce((acc, { kind, label, iri }) => {
      if (!acc[iri.value]) {
        acc[iri.value] = {
          kind,
          labels: [],
          iri,
        };
      }
      acc[iri.value].labels.push({
        value: label.value,
        language: label.language,
      });
      return acc;
    }, {});
    const groupedComponents: GroupedComponents[] = Object.values(componentsByIri);

    this.cachedComponents = groupedComponents
      .reduce((componentsProp: ComponentsCache, { kind, labels, iri }) => {
        switch (kind.value) {
          case "http://purl.org/linked-data/cube#attribute":
            componentsProp.attributes.set(iri.value, new Attribute({ labels, iri }));
            break;
          case "http://purl.org/linked-data/cube#dimension":
            componentsProp.dimensions.set(iri.value, new Dimension({ labels, iri }));
            break;
          case "http://purl.org/linked-data/cube#measure":
            componentsProp.measures.set(iri.value, new Measure({ labels, iri }));
            break;
          default:
            throw new Error(`Unknown component kind ${kind.value}`);
        }
        return componentsProp;
      }, {
        attributes: new Map(),
        dimensions: new Map(),
        measures: new Map(),
      });

    this.componentsLoaded = true;
  }
}
