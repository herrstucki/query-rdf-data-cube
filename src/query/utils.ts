import { literal, namedNode, variable } from "@rdfjs/data-model";
import { Variable } from "rdf-js";
import Component from "../components";
import { ICubeOptions } from "../datacube";
import Binding from "../expressions/binding";
import Operator from "../expressions/operator";
import { ArrayExpr, IExpr, into, isTerm, TermExpr } from "../expressions/utils";
import { BindPattern, BlockPattern, Expression, FilterPattern, OperationExpression, Tuple } from "../sparqljs";

export type PredicateFunction = (data: Selects) => Component;
export type Selects = Record<string, Component>;

export interface IState {
  selects: Selects;
  filters: IExpr[];
  groupBys: Array<PredicateFunction | string>;
  havings: PredicateFunction[];
  offset: number;
  limit: number;
  order: Component[];
}

// tslint:disable-next-line: no-empty-interface
export interface IQueryOptions extends ICubeOptions {}

export const baseState: IState = {
  selects: {},
  filters: [],
  groupBys: [],
  havings: [],
  offset: 0,
  limit: 10,
  order: [],
};

export interface ILangBindings {
  binding: Variable;
  labelBinding: Variable;
  labelLangBinding: Variable;
}

/**
 * Convert [[Operator]] arguments into SPARQL.js `Expression`s
 *
 * @param {Operator} operator
 */
export function operatorArgsToExpressions(args: IExpr[]): Expression[] {
  const expressions = args.map((arg: IExpr): Expression => {
    if (isTerm(arg)) {
      return arg;
    }
    if (arg instanceof Binding) {
      return variable(arg.name);
    }
    if (arg instanceof Operator) {
      return createOperationExpression(arg);
    }
    if (arg instanceof TermExpr) {
      return arg.term;
    }
    if (arg instanceof ArrayExpr) {
      const tuple: Tuple = operatorArgsToExpressions(Array.from(arg.xs).map(into));
      return tuple;
    }
  }).filter((x) => {
    const transformed = Boolean(x);
    if (!transformed) {
      throw new Error("Unrecognized filter argument type");
    }
    return transformed;
  });
  return expressions;
}

export function createOperationExpression(operator: Operator): OperationExpression {
  const operationExpression: OperationExpression = {
    type: "operation",
    operator: operator.operator,
    args: operatorArgsToExpressions(operator.args),
  };
  return operationExpression;
}

/**
 * When `.filter` is called several times, apply a logical AND between all filters.
 */
export function combineFilters(operations: OperationExpression[]): FilterPattern {
  let combined;
  if (operations.length > 1) {
    combined = operations
      .reduce((acc, op) => {
        acc.args.push(op);
        return acc;
      }, {
        operator: "&&",
        type: "operation",
        args: [],
      });
  } else {
    combined = operations[0];
  }
  return {
    type: "filter",
    expression: combined,
  };
}

function langMatchExpression(lang: string, binding: Variable): OperationExpression {
  return {
    type: "operation",
    operator: "langmatches",
    args: [
      {
        type: "operation",
        operator: "lang",
        args: [ binding ],
      },
      literal(lang),
    ],
  };
}

function langExactMatchExpression(lang = "", binding: Variable): OperationExpression {
  return {
    type: "operation",
    operator: "=",
    args: [
      {
        type: "operation",
        operator: "lang",
        args: [ binding ],
      },
      literal(lang),
    ],
  };
}

function langFilter(labelLangBinding: Variable, langs: string[]): FilterPattern {
  let languages = langs.filter(Boolean);
  if (!languages.length) {
    // no lang specified, default to 'empty' lang
    return {
      type: "filter",
      expression: langExactMatchExpression("", labelLangBinding),
    };
  }

  // at least one non-empty lang, add the 'empty' lang as fallback
  languages = langs.concat("");
  // we now have at least two langs to work with
  const lang1 = languages.shift();
  const lang2 = languages.shift();

  let expression: OperationExpression = {
    type: "operation",
    operator: "||",
    args: [
      lang1 === "" ? langExactMatchExpression(lang1, labelLangBinding) : langMatchExpression(lang1, labelLangBinding),
      lang2 === "" ? langExactMatchExpression(lang2, labelLangBinding) : langMatchExpression(lang2, labelLangBinding),
    ],
  };

  // remove the first lang until there's no lang left
  let extraLang = languages.shift();
  while (typeof extraLang !== "undefined") {
    // nest 'OR' operations
    expression = {
      type: "operation",
      operator: "||",
      args: [
        expression,
        extraLang === ""
          ? langExactMatchExpression("", labelLangBinding)
          : langMatchExpression(extraLang, labelLangBinding),
      ],
    };
    extraLang = languages.shift();
  }
  return {
    type: "filter",
    expression,
  };
}

export function generateLangOptional(bindings: ILangBindings, langs: string[]): BlockPattern {
  const {binding, labelBinding, labelLangBinding} = bindings;
  const findLabel: BlockPattern = {
    type: "optional",
    patterns: [
      {
        type: "bgp",
        triples: [
          {
            subject: binding,
            predicate: {
              type: "path",
              pathType: "|",
              items: [
                namedNode("http://www.w3.org/2000/01/rdf-schema#label"),
                namedNode("http://www.w3.org/2004/02/skos/core#prefLabel"),
              ],
            },
            object: labelLangBinding,
          },
        ],
      },
      langFilter(labelLangBinding, langs),
    ],
  };

  return findLabel;
}

export function generateLangCoalesce(bindings: ILangBindings, langs: string[]): BindPattern {
  const {binding, labelBinding, labelLangBinding} = bindings;
  const coalesceLabel: BindPattern = {
    type: "bind",
    variable: labelBinding,
    expression: {
      type: "operation",
      operator: "coalesce",
      args: [
        labelLangBinding,
        literal(""),
      ],
    },
  };

  return coalesceLabel;
}
