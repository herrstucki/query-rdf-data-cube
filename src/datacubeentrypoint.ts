import { namedNode, variable } from "@rdfjs/data-model";
import { NamedNode } from "rdf-js";
import { Generator as SparqlGenerator } from "sparqljs";
import DataSet, { DataSetOptions, Label } from "./dataset";
import { generateLangCoalesce, generateLangOptionals, prefixes } from "./query/utils";
import SparqlFetcher, { SparqlFetcherOptions } from "./sparqlfetcher";
import { SelectQuery } from "./sparqljs";

export interface EntryPointOptions {
  languages?: string[];
  fetcher?: SparqlFetcherOptions;
}

export type SerializedDataCubeEntryPoint = {
  endpoint: string,
  languages: string[],
  datasets: string[],
};

export default class DataCubeEntryPoint {
  /**
   * Deserializes a DataCubeEntryPoint from JSON generated by DataCubeEntryPoint#toJSON
   */
  public static fromJSON(json: string): DataCubeEntryPoint {
    const obj: SerializedDataCubeEntryPoint = JSON.parse(json);
    const entryPoint = new DataCubeEntryPoint(obj.endpoint, {
      languages: obj.languages,
    });
    entryPoint.cachedDatasets = obj.datasets.reduce((map, str) => {
      const dataset: DataSet = DataSet.fromJSON(str);
      map.set(dataset.iri, dataset);
      return map;
    }, new Map());
    return entryPoint;
  }

  private endpoint: string;
  private languages: string[];
  private fetcher: SparqlFetcher;
  private cachedDatasets: Map<string, DataSet>;
  private allDatasetsLoaded: boolean = false;
  private cachedGraphs: NamedNode[];
  private graphsLoaded: boolean = false;

  /**
   * A DataCubeEntryPoint queries a SPARQL endpoint and retrieves [[DataSet]]s and
   * their [[Dimension]]s, [[Measure]]s and [[Attribute]]s.
   * @class DataCubeEntryPoint
   * @param endpoint SPARQL endpoint
   * @param options Options
   * @param options.languages Languages in which to get the labels, by priority, e.g. `["de", "en"]`.
   * Passed down to [[DataSet]]s and [[DataSetQuery]].
   */
  constructor(endpoint: string, options: EntryPointOptions = {}) {
    this.endpoint = endpoint;
    this.languages = options.languages || [];
    this.fetcher = new SparqlFetcher(endpoint, options.fetcher || {});
    this.cachedDatasets = new Map();
    this.cachedGraphs = [];
  }

  /**
   * Serializes a DataCubeEntryPoint to JSON in a way that makes it deserializable
   * by calling DataCubeEntryPoint#fromJSON
   */
  public toJSON(): string {
    const obj: SerializedDataCubeEntryPoint = {
      endpoint: this.endpoint,
      languages: this.languages,
      datasets: Array.from(this.cachedDatasets.values()).map((dataset) => dataset.toJSON()),
    };
    return JSON.stringify(obj);
  }

  /**
   * Fetch all [[DataSet]]s from the endpoint.
   */
  public async datasets(): Promise<DataSet[]> {
    if (this.allDatasetsLoaded) {
      return Array.from(this.cachedDatasets.values());
    }

    const sparql = this.generateQuery();
    const queryResult = await this.fetcher.select(sparql);
    this.cacheDatasets(queryResult);
    this.allDatasetsLoaded = true;
    return Array.from(this.cachedDatasets.values());
  }

  /**
   * Fetch a [[DataSet]] by its IRI.
   *
   * @param iri IRI of the DataSet to return.
   */
  public async datasetByIri(dataSetIri: string): Promise<DataSet> {
    const found = Array.from(this.cachedDatasets.values()).find((dataset) => dataset.iri === dataSetIri);
    if (found) {
      return found;
    }
    const sparql = this.generateQuery({ dataSetIri: namedNode(dataSetIri) });
    const queryResult = await this.fetcher.select(sparql);
    if (!queryResult.length) {
      throw new Error(`No dataset with iri <${dataSetIri}> on ${this.endpoint}`);
    }
    this.cacheDatasets(queryResult);
    return this.datasetByIri(dataSetIri);
  }

  /**
   * Fetch [[DataSet]]s by their graph IRI.
   *
   * @param graphIri IRI of the graph to look for in all DataSets.
   */
  public async datasetsByGraphIri(iri: string): Promise<DataSet[]> {
    const datasets = Array.from(this.cachedDatasets.values())
      .filter((ds) => ds.graphIri === iri);
    if (datasets.length) {
      return datasets;
    }
    const graphIri = namedNode(iri);

    const sparql = this.generateQuery({ graphIri });
    const queryResult = await this.fetcher.select(sparql);
    if (!queryResult.length) {
      // avoid infinite recursion
      throw new Error(`Cannot find graph <${iri}> in ${this.endpoint}`);
    }
    this.cacheDatasets(queryResult.map((result) => {
      result.graphIri = graphIri;
      return result;
    }));
    return this.datasetsByGraphIri(iri);
  }

  /**
   * Fetch all graphs from the endpoint.
   */
  public async graphs(): Promise<NamedNode[]> {
    if (this.graphsLoaded) {
      return this.cachedGraphs;
    }
    const query = `
      PREFIX qb: <http://purl.org/linked-data/cube#>
      SELECT DISTINCT ?graph WHERE {
        GRAPH ?graph {
          ?dataset a qb:DataSet .
        }
      }
    `;
    const graphs = await this.fetcher.select(query);
    this.graphsLoaded = true;
    return this.cachedGraphs = graphs.map(({ graph }) => graph);
  }

  private cacheDatasets(datasets: Array<{ iri: NamedNode, label: Label, graphIri: string }>) {
    const datasetsByIri = datasets.reduce((acc, { iri, label, graphIri }) => {
      if (!acc[iri.value]) {
        acc[iri.value] = {
          iri,
          labels: [],
          graphIri,
          languages: this.languages,
        };
      }
      acc[iri.value].labels.push({
        value: label.value,
        language: label.language,
      });
      return acc;
    }, {});

    Object.entries(datasetsByIri)
      .forEach(([iri, dataset]: [string, DataSetOptions]) => {
        this.cachedDatasets.set(iri, new DataSet(this.endpoint, dataset));
      });
  }

  private generateQuery({graphIri, dataSetIri}: {graphIri?: NamedNode, dataSetIri?: NamedNode} = {}) {
    const graphIriBinding = variable("graphIri");
    const iriBinding = variable("iri");
    const labelBinding = variable("label");

    const query: SelectQuery = {
      prefixes,
      queryType: "SELECT",
      variables: [
        iriBinding,
        graphIriBinding,
        labelBinding,
      ],
      where: [
        {
          type: "graph",
          name: graphIri || graphIriBinding,
          patterns: [
            {
              type: "bgp",
              triples: [
                {
                  subject: iriBinding,
                  predicate: namedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"),
                  object: namedNode("http://purl.org/linked-data/cube#DataSet"),
                },
              ],
            },
            ...generateLangOptionals(iriBinding, labelBinding, this.languages),
          ],
        },
        generateLangCoalesce(labelBinding, this.languages),
      ],
      type: "query",
    };
    if (dataSetIri) {
      query.where.push({
        type: "filter",
        expression: {
          type: "operation",
          operator: "=",
          args: [
            iriBinding,
            dataSetIri,
          ],
        },
      });
    }

    const generator = new SparqlGenerator({ allPrefixes: true });
    return generator.stringify(query);
  }
}
