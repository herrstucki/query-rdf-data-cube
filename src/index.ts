import Attribute from "./components/attribute";
import { SerializedComponent } from "./components/component";
import Dimension from "./components/dimension";
import Measure from "./components/measure";

import BaseExpr from "./expressions/base";
import Binding from "./expressions/binding";
import Operator from "./expressions/operator";
import { ArrayExpr, IExpr, into, IntoExpr, isLiteral, isTerm, TermExpr, toLiteral } from "./expressions";

import DataSetQuery from "./query/datasetquery";
import { baseState, IQueryOptions, IState, PredicateFunction, Selects } from "./query/utils";
import { combineFilters, createOperationExpression, operatorArgsToExpressions } from "./query/utils";
import { generateLangCoalesce, generateLangOptionals, prefixes } from "./query/utils";

import DataCubeEntryPoint, { EntryPointOptions, SerializedDataCubeEntryPoint } from "./datacubeentrypoint";
import DataSet, { Label } from "./dataset";
import SparqlFetcher, { SparqlFetcherOptions } from "./sparqlfetcher";

export default DataCubeEntryPoint;
export {
  ArrayExpr,
  Attribute,
  BaseExpr,
  Binding,
  DataCubeEntryPoint,
  DataSet,
  DataSetQuery,
  Dimension,
  EntryPointOptions,
  IExpr,
  IQueryOptions,
  SparqlFetcherOptions,
  IState,
  IntoExpr,
  Label,
  Measure,
  Operator,
  PredicateFunction,
  Selects,
  SerializedComponent,
  SerializedDataCubeEntryPoint,
  SparqlFetcher,
  TermExpr,
  baseState,
  combineFilters,
  createOperationExpression,
  generateLangCoalesce,
  generateLangOptionals,
  into,
  isLiteral,
  isTerm,
  operatorArgsToExpressions,
  prefixes,
  toLiteral,
};
