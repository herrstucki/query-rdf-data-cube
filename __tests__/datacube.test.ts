import { DataCubeEntryPoint, EntryPointOptions } from "../src/entrypoint";
import { fetch } from "./utils/fetch-mock";

const newCube = (endpoint: string, languages?: string[]) => {
  const options: EntryPointOptions = {
    fetcher: {
      fetch,
    },
  };
  if (languages) {
    options.languages = languages;
  }
  return new DataCubeEntryPoint(endpoint, options);
};

describe("dataCube", () => {
  it("gets component values", async () => {
    const entryPoint = newCube("https://trifid-lindas.test.cluster.ldbar.ch/query");
    const dataCubes = await entryPoint.dataCubes();
    const dataCube = dataCubes[0];

    const dimensions = await dataCube.dimensions();

    const sizeClasses = dimensions[1];

    const values = (await dataCube.componentValues(sizeClasses))
      .filter((value) => value.label.value !== "50 - 100 ha");

    const values2 = await dataCube.query()
      .select({ size: sizeClasses })
      .filter(({ size }) => size.notEquals("50 - 100 ha"))
      .componentValues();

    expect(values).toEqual(values2);
  });

  it("gets component values min/max", async () => {
    const entryPoint = newCube("https://ld.stadt-zuerich.ch/query");
    const dataCubes = await entryPoint.dataCubes();
    const dataCube = dataCubes.find((cube) => cube.iri.endsWith("BEW-RAUM-ZEIT"));

    const dimensions = await dataCube.dimensions();
    const time = dimensions.find((dimension) => dimension.iri.value.endsWith("/ZEIT"));
    const timeMinMax = await dataCube.componentMinMax(time);

    expect(Object.keys(timeMinMax)).toEqual(["min", "max"]);
    expect(parseInt(timeMinMax.min.value, 10)).toBeLessThan(parseInt(timeMinMax.max.value, 10));

    Object.values(timeMinMax).forEach((value) => {
      expect(value.termType).toBe("Literal");
    });
  });
});
