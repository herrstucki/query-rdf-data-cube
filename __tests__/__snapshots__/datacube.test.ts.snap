// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataCube .componentValues() 1`] = `
Array [
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/0",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/1",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/2",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/3",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/4",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/5",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/6",
    },
  },
  Object {
    "label": Literal {
      "datatype": NamedNode {
        "value": "http://www.w3.org/2001/XMLSchema#string",
      },
      "language": "",
      "value": "",
    },
    "value": NamedNode {
      "value": "http://example.org/anzahl-forstbetriebe/property/1/7",
    },
  },
]
`;
